### PR TITLE
RAJA Local Array

### PR DESCRIPTION
**Pull request by artv3**

### CODE REVIEW

**Comment by artv3** [comment]
@ajkunen , @trws , I have something functional now for shared memory. Here, I introduced a new statement CreateSharedMemory which allocates memory at a user specified location. This enables shared memory to be thread private or a scratch pad depending where its located. A thought I had was to perhaps consider using memory on the stack for this type of computation. At the moment its a proof of concept but it has the bones for enabling new types of computing patterns.

**Comment by rhornung67** [comment]
@ajkunen would it be a good idea to have @artv3 look into developing a static permutation capability to the shared memory creation part in this branch?

File: examples/mat-transpose.cpp [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+  RAJA::kernel_param<KERNEL_EXEC_POL>(iSpace,
+                                      RAJA::make_tuple(myTile, myTile2),
+
+      [=] (int tx, int ty, int bx, int by, mySharedMemory &myTile,  mySharedMemory &myTile2) {
+
+           int col = bx * TILE_DIM + tx;  // Matrix column index
+           int row = by * TILE_DIM + ty;  // Matrix row index
+          if(row < N_rows and col < N_cols){
+            (*myTile.SharedMem)(ty,tx)  = Aview(row, col);
```
Comment (ajkunen) [inline examples/mat-transpose.cpp]: @artv3 How can we do this without the explicit tile/index calculation?

Comment (artv3) [reply examples/mat-transpose.cpp]: We can merge the initialization step of CreateShmem into SetShmemWindow. I would have to re-work some of the internals in SetShmemWindow to have it work with the shared memory wrapper, but should be straight forward.

**Comment by artv3** [comment]
@ajkunen We can merge the initialization step of CreateShmem into SetShmemWindow. I would have to re-work some of the internals in SetShmemWindow to have it work with the shared memory wrapper, but should be straight forward.

File: test/unit/test-sharedmem.cpp [hunk 2]
```diff
@@ -0,0 +1,929 @@
...
+  for(int r=0; r<N; ++r){
+    for(int c=0; c<P; ++c){
+      int dot = 0.0;
+      for(int k=0; k<M; ++k){
+        dot += Aview(r,k)*Bview(k,c);
+      }
+      C_solView(r,c) = dot;
+    }
+  }
+
```
Comment (artv3) [inline test/unit/test-sharedmem.cpp]: @trws , I put together a new iteration of the shared memory framework. Below is a matrix-multiplication example which exercises new RAJA memory objects. In particular, I introduce RAJA memory objects which may be paired with one of the following templates:
RAJA::cpu_tile_mem  - allocates memory on the CPU stack
RAJA::cuda_priv_mem – allocates thread private memory
RAJA::cuda_shared_mem – allocates CUDA shared memory

Furthermore, I introduced a new statement called CreateShmem (name is a work in progress) which enables users to allocate memory in between loops. This enables us to emulate shared memory on the CPU. I was wondering if you might have some time to glance at the work. The code is compatible with a variety of different execution policies as well.

**Comment by artv3** [comment]
I did some clean up, and made changes based on Tuesday's discussion. I think this branch is ready for a review. If you could please take a look and let me know what you think? @rhornung67 , @ajkunen , @trws . Thank you for your time. 

Also @ajkunen , regarding projections for the static layout, it looks like it was just naturally supported. So no changes were necessary.

File: include/RAJA/util/ShmemTile.hpp [hunk 3]
```diff
@@ -38,6 +38,72 @@
 namespace RAJA
 {
 
+//Policies for RAJA scoped arrays
+struct cpu_tile_mem;
+struct cuda_thread_mem;
+struct cuda_shared_mem;
+
+template<camp::idx_t ... Sizes>
+using param_idx = camp::idx_seq<Sizes...>;
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply include/RAJA/util/ShmemTile.hpp]: @ajkunen good catch!

File: test/unit/test-sharedmem.cpp [hunk 4]
```diff
@@ -0,0 +1,1032 @@
...
+  RAJA::View<double, RAJA::Layout<DIM>> Aview(A, N_rows, N_cols);
+  RAJA::View<double, RAJA::Layout<DIM>> Bview(B, N_rows, N_cols);
+
+  for (int row = 0; row < N_rows; ++row) {
+    for (int col= 0 ; col < N_cols; ++col) {
+      Aview(row, col) = col;
+    }
+  }
+
+  using SharedTile = TypedLocalArray<double, RAJA::SizeList<TILE_DIM,TILE_DIM>, TY, TX>;
```
Comment (rhornung67) [inline test/unit/test-sharedmem.cpp]: @artv3 here, you pass strongly-typed index args as template params to define the tile type, but then below you have 'int' args passed to the lambda and you manually cast them back to strongly-typed. Can't you have strongly-typed index args passed to the lambda?

File: test/unit/test-sharedmem.cpp [hunk 5]
```diff
@@ -0,0 +1,1032 @@
...
+  SharedTile myTile, myTile2;
+
+  RAJA::kernel_param<Pol>(RAJA::make_tuple(RAJA::RangeSegment(0, inner_Dim0), RAJA::RangeSegment(0,inner_Dim1),
+                                           RAJA::RangeSegment(0, outer_Dim0), RAJA::RangeSegment(0,outer_Dim1)),
+                          RAJA::make_tuple(myTile, myTile2),
+
+  //Load data into shared memory
+  [=] RAJA_HOST_DEVICE (int tx, int ty, int bx, int by, SharedTile &myTile, SharedTile &) {
```
Comment (rhornung67) [inline test/unit/test-sharedmem.cpp]: Here...

File: test/unit/test-sharedmem.cpp (hunk 4 above)
Comment (artv3) [reply test/unit/test-sharedmem.cpp]: That's a good point, I just updated it.

**Comment by artv3** [comment]
@rhornung67 , @ajkunen , I just made a few more fixes. Could I get another review?

**Pull request merged**